freight/providers/pipeline.py

```diff
@@ -0,0 +1,359 @@
+__all__ = ["PipelineProvider"]
+
+import sys
+from time import sleep, time
+from typing import Optional, List, Callable, Tuple
+from functools import partial
+from dataclasses import dataclass, asdict
+
+from kubernetes import client
+from kubernetes.config import new_client_from_config, ConfigException
+
+from .base import Provider
+from freight.utils.workspace import Workspace
+from freight.models import App, Deploy
+
+
+@dataclass(frozen=True)
+class TaskContext:
+    id: str
+    date_created: str
+    name: str
+    environment: str
+    sha: str
+    prev_sha: str
+    ref: str
+
+
+@dataclass(frozen=True)
+class KubernetesContext:
+    client: client.ApiClient
+
+
+@dataclass(frozen=True)
+class PipelineContext:
+    task: TaskContext
+    kube: Optional[KubernetesContext]
+    workspace: Workspace
+
+
+class StepFailed(Exception):
+    pass
+
+
+class PipelineProvider(Provider):
+    def get_options(self):
+        return {
+            "steps": {"required": True, "type": list},
+            "kubernetes": {"required": False, "type": dict},
+        }
+
+    def get_config(self, workspace, task) -> dict:
+        # from yaml import safe_load
+        # TODO(mattrobenolt): Load and merge config (.freight.yml) from
+        # within the workspace, so a config could be left out entirely.
+        return {**task.provider_config, **{}}
+
+    def execute(self, workspace, task):
+        deploy = Deploy.query.filter(Deploy.task_id == task.id).first()
+        return self.execute_deploy(workspace, deploy, task)
+
+    def execute_deploy(self, workspace, deploy, task):
+        app = App.query.get(task.app_id)
+        prev_sha = app.get_previous_sha(deploy.environment, current_sha=task.sha)
+
+        config = self.get_config(workspace, task)
+        if config["kubernetes"]:
+            kube_context = KubernetesContext(
+                load_kube_credentials(config["kubernetes"])
+            )
+        else:
+            kube_context = None
+
+        task_context = TaskContext(
+            str(deploy.id),
+            str(task.date_created),
+            app.name,
+            deploy.environment,
+            task.sha,
+            prev_sha or "",
+            task.ref,
+        )
+
+        context = PipelineContext(
+            task=task_context, kube=kube_context, workspace=workspace
+        )
+
+        for i, step in enumerate(config["steps"]):
+            workspace.log.info(f"Running Step {i+1} ({step['kind']})")
+            run_step(step, context)
+            workspace.log.info(f"Finished Step {i+1}")
+
+
+def load_kube_credentials(config: dict) -> client.ApiClient:
+    try:
+        return new_client_from_config(context=config["context"])
+    except KeyError:
+        pass
+
+    try:
+        credentials = config["credentials"]
+    except KeyError:
+        return
+
+    assert credentials["kind"] == "gcloud"
```

Comment:
Are assertions fairly standard in freight? To me this might be a bit more clear as an exception.

Reply:
No, but this is ultimately a configuration error. And our validation of the config when it's saved is not robust enough to support deep validation like we'd need here, and I don't wanna write that now. An assert here will just cause the build to fail with an AssertionError, which frankly, is just identical to me wrapping it in a different error.